src/System.Private.CoreLib/src/System/StartupHookProvider.cs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+using Internal.IO;
+
+namespace System
+{
+    internal class StartupHookProvider
+    {
+        private static string initializeMethodName = "Initialize";
+
+        // Parse a string specifying a list of assemblies and types
+        // containing a startup hook, and call each hook in turn.
+        private static void ProcessStartupHooks(IntPtr startupHooksVariable)
+        {
+            Debug.Assert(startupHooksVariable != IntPtr.Zero);
+            string[] startupHooks = Marshal.PtrToStringUTF8(startupHooksVariable).Split(new char[] { Path.PathSeparator });
+
+            // Parse startup hook variable
+            List<Tuple<string, string>> startupHooksParsed = new List<Tuple<string, string>>();
+            foreach (string startupHook in startupHooks)
+            {
+                string[] assemblyPathAndTypeName = startupHook.Split(new char[] { '!' });
+                if (assemblyPathAndTypeName.Length != 2)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                string assemblyPath = assemblyPathAndTypeName[0];
+                string typeName = assemblyPathAndTypeName[1];
+                if (String.IsNullOrEmpty(assemblyPath) || String.IsNullOrEmpty(typeName))
+                {
+                    throw new ArgumentException(SR.Argument_InvalidStartupHookSyntax);
+                }
+                startupHooksParsed.Add(Tuple.Create(assemblyPath, typeName));
+            }
+
+            // Ensure the startup dlls exist
+            foreach (var startupHook in startupHooksParsed)
+            {
+                var assemblyPath = startupHook.Item1;
+                if (!File.Exists(assemblyPath))
```

Comment:
This is kind of a lie. I say that because we check for existence against what the user supplied - yay, but then in `CallStartupHook()`, we call `Path.GetFullPath()` and then use that path. It might be better to resolve to the full path here and report that.

Reply:
My thinking is that it makes sense for the reported missing path to be the user-supplied one, for easy comparison to the startup hook variable. GetFullPath is an implementation detail (it's a requirement for LoadFromAssemblyPath). As I understand it, they will point to the same file on disk (unless the filesystem changes... in which case LoadFromAssemblyPath will throw a similar error for the full path).

On the other hand, I can see an argument for reporting the full path - it might be easier to find out where this is actually looking if the working directory isn't known for some reason. Is this your concern?

Reply:
I see, thanks! I'll take you recommendation.

Reply:
Hmm, there's not an existing exception string (that I could find) that specifies both the given and full paths. It would be no trouble to add one, but I'd prefer consistency across similar file not found errors in the runtime.

As part of the ongoing work to improve assembly load diagnostics, we should look at all uses of this and other resource strings in FileNotFoundException (and related exceptions), and consider whether we want to introduce better or more specific error reporting for all of these cases together.

For now I think I'll throw an exception with the message `"Could not resolve assembly '{fullPath}'"`, using the existing `FileNotFound_ResolveAssembly` resource string. This way they'll at least see the full path.

Unless you have a strong preference for introducing the verbose error now?